### PR TITLE
Add ingoing and outgoing counts to processorhelper

### PR DESCRIPTION
**Pull request by djaglowski**
#### Description

Implements ingoing and outgoing counts as described in https://github.com/open-telemetry/opentelemetry-collector/issues/10708.

### CODE REVIEW

**Review by iblancasa — APPROVED** [review]

**Comment by djaglowski** [comment]
Any thoughts on this, @open-telemetry/collector-contrib-approvers?

File: processor/processorhelper/logs.go [hunk 1]
```diff
@@ -39,12 +40,25 @@ func NewLogsProcessor(
 		return nil, errors.New("nil logsFunc")
 	}
 
+	if set.MeterProvider == nil {
+		set.MeterProvider = noop.NewMeterProvider()
```
Comment (codeboten) [inline processor/processorhelper/logs.go:44]: i wouldn't expect this to be needed here, can you confirm whether it is?

File: processor/processorhelper/metrics.go [hunk 2]
```diff
@@ -39,12 +40,25 @@ func NewMetricsProcessor(
 		return nil, errors.New("nil metricsFunc")
 	}
 
+	if set.MeterProvider == nil {
+		set.MeterProvider = noop.NewMeterProvider()
```
Comment (codeboten) [inline processor/processorhelper/metrics.go:44]: same question as above

File: processor/processorhelper/metrics_test.go [hunk 3]
```diff
@@ -67,3 +73,72 @@ func newTestMProcessor(retError error) ProcessMetricsFunc {
 		return md, retError
 	}
 }
+
+func TestMetricsProcessor_RecordInOut(t *testing.T) {
+	// Regardless of how many logs are ingested, emit just one
```
Comment (codeboten) [inline processor/processorhelper/metrics_test.go:78]:
```suggestion
	// Regardless of how many data points are ingested, emit just one
```

File: processor/processorhelper/traces_test.go [hunk 4]
```diff
@@ -67,3 +73,72 @@ func newTestTProcessor(retError error) ProcessTracesFunc {
 		return td, retError
 	}
 }
+
+func TestTracesProcessor_RecordInOut(t *testing.T) {
+	// Regardless of how many logs are ingested, emit just one
```
Comment (codeboten) [inline processor/processorhelper/traces_test.go:78]:
```suggestion
	// Regardless of how many spans are ingested, emit just one
```

File: processor/processorhelper/logs.go (hunk 1 above)
Comment (djaglowski) [reply processor/processorhelper/logs.go]: You're right, it was a problem with a particular test, not something that needs to be addressed here.

**Review by djaglowski — COMMENTED** [review]
I've rebased and resolved the conflict.

File: processor/processorhelper/metrics.go (hunk 2 above)
Comment (djaglowski) [reply processor/processorhelper/metrics.go]: Removed for all three signals.

**Comment by djaglowski** [comment]
A contrib test [failed](https://github.com/open-telemetry/opentelemetry-collector/actions/runs/10743816191/job/29799409203?pr=10910#step:4:707), but I think it expected with this change:
```
=== Failed
=== FAIL: . TestFilterMetricProcessorTelemetry (0.00s)
make[3]: *** [../../Makefile.Common:128: test] Error 1
    generated_component_telemetry_test.go:55: 
        	Error Trace:	/tmp/opentelemetry-collector-contrib/processor/filterprocessor/generated_component_telemetry_test.go:55
        	            				/tmp/opentelemetry-collector-contrib/processor/filterprocessor/metrics_test.go:423
        	Error:      	Not equal: 
make[2]: *** [Makefile:187: processor/filterprocessor] Error 2
make[1]: *** [Makefile:127: gotest] Error 2
make: *** [Makefile:260: check-contrib] Error 2
        	            	expected: 1
        	            	actual  : 3
        	Test:       	TestFilterMetricProcessorTelemetry

=== FAIL: . TestFilterMetricProcessorTelemetry (re-run 1) (0.00s)
    generated_component_telemetry_test.go:55: 
        	Error Trace:	/tmp/opentelemetry-collector-contrib/processor/filterprocessor/generated_component_telemetry_test.go:55
        	            				/tmp/opentelemetry-collector-contrib/processor/filterprocessor/metrics_test.go:423
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 3
        	Test:       	TestFilterMetricProcessorTelemetry
```

**Comment by codeboten** [comment]
> A contrib test [failed](https://github.com/open-telemetry/opentelemetry-collector/actions/runs/10743816191/job/29799409203?pr=10910#step:4:707), but I think it expected with this change:
> 
> ```
> === Failed
> === FAIL: . TestFilterMetricProcessorTelemetry (0.00s)
> make[3]: *** [../../Makefile.Common:128: test] Error 1
>     generated_component_telemetry_test.go:55: 
>         	Error Trace:	/tmp/opentelemetry-collector-contrib/processor/filterprocessor/generated_component_telemetry_test.go:55
>         	            				/tmp/opentelemetry-collector-contrib/processor/filterprocessor/metrics_test.go:423
>         	Error:      	Not equal: 
> make[2]: *** [Makefile:187: processor/filterprocessor] Error 2
> make[1]: *** [Makefile:127: gotest] Error 2
> make: *** [Makefile:260: check-contrib] Error 2
>         	            	expected: 1
>         	            	actual  : 3
>         	Test:       	TestFilterMetricProcessorTelemetry
> 
> === FAIL: . TestFilterMetricProcessorTelemetry (re-run 1) (0.00s)
>     generated_component_telemetry_test.go:55: 
>         	Error Trace:	/tmp/opentelemetry-collector-contrib/processor/filterprocessor/generated_component_telemetry_test.go:55
>         	            				/tmp/opentelemetry-collector-contrib/processor/filterprocessor/metrics_test.go:423
>         	Error:      	Not equal: 
>         	            	expected: 1
>         	            	actual  : 3
>         	Test:       	TestFilterMetricProcessorTelemetry
> ```

The test is likely a failure of the new metrics being emitted and the original test not expecting them to be there. Please open a PR to update the test as a follow up

**Comment by djaglowski** [comment]
Thanks for the review @codeboten. I created https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35073 to update the contrib test.

File: processor/processorhelper/metrics_test.go [hunk 5]
```diff
@@ -67,3 +74,78 @@ func newTestMProcessor(retError error) ProcessMetricsFunc {
...
+
+	incomingMetrics := pmetric.NewMetrics()
+	dps := incomingMetrics.ResourceMetrics().AppendEmpty().ScopeMetrics().AppendEmpty().Metrics().AppendEmpty().SetEmptySum().DataPoints()
+
+	// Add 3 data points to the incoming
+	dps.AppendEmpty()
+	dps.AppendEmpty()
+	dps.AppendEmpty()
+
+	metricReader := sdkmetric.NewManualReader()
```
Comment (codeboten) [inline processor/processorhelper/metrics_test.go:94]: Could the test code here use the generated test code? Can be done in a follow up PR

**Review by codeboten — APPROVED** [review]

**Pull request merged**